storage: Anaconda mode

mvollmer opened the pull request:
Anaconda code: https://github.com/mvollmer/anaconda-webui/tree/cockpit-storage

Demo: https://www.youtube.com/watch?v=x3KfKJhBxD0

Next steps after this PR:

 - Discourage the user from actually mounting new filesystems by only offering "Format only" in the format dialog.

mvollmer commented:
@jkonecny12, please check this out!

M4rtinK commented on doc/anaconda.md:
```diff
@@ -0,0 +1,73 @@
+However, Cockpit (via UDisks2) will still write the new mount point
+configuration into the real /etc/fstab (_not_ /sysroot/etc/fstab).
+
+Anaconda will read the real /etc/fstab and reverse engineer the
```
@vojtechtrefny @VladimirSlavik @poncovka I think Blivet has some support for fstab parsing, right ? I just wonder what else could be there in fstab & if we can't get into a situation where we can't tell what we want to forward to the target system & what now.

vojtechtrefny replied:
Blivet doesn't have support for parsing fstab yet, but it is planned. But I think it would be better to not write to fstab -- in the "anaconda" mode Cockpit could simply write the mount points to a custom configuration file, all we need in Anaconda is a simple device -> mountpoint "mapping", we don't need the full fstab entry. We could avoid a lot of potential issues by not writing to fstab on the live image (which is not empty by the way).

vojtechtrefny replied:
We (now speaking as a udisks developer) could also add a new option/type to `AddConfigurationItem` to be able to write to a fstab in a different location than the default `/etc/fstab`.

mvollmer replied:
> We (now speaking as a udisks developer) could also add a new option/type to [AddConfigurationItem](http://storaged.org/doc/udisks2-api/latest/gdbus-org.freedesktop.UDisks2.Block.html#gdbus-method-org-freedesktop-UDisks2-Block.AddConfigurationItem) to be able to write to a fstab in a different location than the default `/etc/fstab`.

I think that would only help if UDisks2 would also watch and read that different location. If modifying the real /etc/fstab is not acceptable, I think I would rather change Cockpit Storage to maintain its own "virtual fstab" (in Browser storage?) and use that instead of the UDisks2 API.

mvollmer replied:
> We could avoid a lot of potential issues by not writing to fstab on the live image (which is not empty by the way).

What issues do you see?

I can't think of any, tbh.  Cockpit Storage will of course not corrupt /etc/fstab, if that is one of your concerns.

mvollmer replied:
> I just wonder what else could be there in fstab & if we can't get into a situation where we can't tell what we want to forward to the target system & what now.

The argument for writing into /etc/fstab is that this needs no code changes to Cockpit Storage. The main challenge I see is that Anaconda needs more code. :-)

There must be Python code for parsing fstab files, but Anaconda probably also needs to look into udev to find the block device that holds a filesystem with a given UUID.

If that is too problematic, Cockpit can also, in addition to /etc/fstab, write a separate file with new mountpoints. Then /etc/fstab would only be changed to make live easier for Cockpit.

mvollmer replied:
> all we need in Anaconda is a simple device -> mountpoint "mapping"

Can the "simple device" also be a device mapper device, like "/dev/dm-2"? Can Anaconda figure out from there that this is a LVM2 logical volume, for example?

mvollmer replied:
Cockpit now stores a mountpoint -> device map for Anaconda.

vojtechtrefny replied:
> I think that would only help if UDisks2 would also watch and read that different location.

I guess we could do that, but that would be something too anaconda-specific and I am not sure if @tbzatek would be happy about it.

> What issues do you see?

My main issue is with the existing entries in fstab and potential for Anaconda to mix these up with the "installer" entries created by Cockpit.

I am also not sure how well Cockpit (or UDisks to be more precise) handles btrfs subvolumes in fstab, but that's a different story.

> There must be Python code for parsing fstab files

Anaconda has some functions for parsing fstab and we are working on more generic support in blivet (see https://github.com/storaged-project/blivet/pull/1119)

> but Anaconda probably also needs to look into udev to find the block device that holds a filesystem with a given UUID.

Blivet has this covered.

> Can the "simple device" also be a device mapper device, like "/dev/dm-2"? Can Anaconda figure out from there that this is a LVM2 logical volume, for example?

Yes, we (blivet) can identify a device by name, uuid, all the `/dev` symlinks, basically everything.

```
In [2]: b.devicetree.resolve_device("/dev/dm-2")
Out[2]: 
LVMLogicalVolumeDevice instance (0x7f53ec3957d0) --
  name = fedora_aida-root  status = True  id = 78
  ...
```

mvollmer replied:
> > I think that would only help if UDisks2 would also watch and read that different location.
> 
> I guess we could do that, but that would be something too anaconda-specific and I am not sure if @tbzatek would be happy about it.
> 
> > What issues do you see?
> 
> My main issue is with the existing entries in fstab and potential for Anaconda to mix these up with the "installer" entries created by Cockpit.

Does Anaconda already look into fstab, or are you referring to Anaconda getting confused during my proposed 'reverse engineering'? If the latter, I think this becomes moot now that Cockpit provides Anaconda with a separate mount-point -> device mapping.

> I am also not sure how well Cockpit (or UDisks to be more precise) handles btrfs subvolumes in fstab, but that's a different story.

Once Cockpit supports btrfs, it will handle the subvolumes correctly, see #16408. But currently it is indeed pretty bad at it.

> > Can the "simple device" also be a device mapper device, like "/dev/dm-2"? Can Anaconda figure out from there that this is a LVM2 logical volume, for example?
> 
> Yes, we (blivet) can identify a device by name, uuid, all the `/dev` symlinks, basically everything.

Nice!

tbzatek replied:
> > I think that would only help if UDisks2 would also watch and read that different location.
> 
> I guess we could do that, but that would be something too anaconda-specific and I am not sure if @tbzatek would be happy about it.

@tbzatek would not be so happy with this I'm afraid.

> > What issues do you see?
> 
> My main issue is with the existing entries in fstab and potential for Anaconda to mix these up with the "installer" entries created by Cockpit.

I have no idea how entries from multiple sources will be visualized (on the D-Bus interface). Besides, the whole RFE seems a little off the UDisks project target.

However, if something like this is about to be needed, we need to think in bigger picture. There are various files that we watch and/or modify. In some cases where these files are consumed by other libraries we pull in - a problem with different prefix in the build time sometimes causes unexpected issues.

Which brings me into another question - should this be considered with (not actually planned yet) mount namespace support? Would this be interesting for your use case?

Feel free to detach this conversation to a separate ticket.

M4rtinK commented on doc/anaconda.md:
```diff
@@ -0,0 +1,73 @@
+storaged is in a special "Anaconda mode" and behaves significantly
+different.
+
+In essence, the storaged page restricts itself to working with the
```
While most real world Live installations will indeed happen from USB stick, I just wanted to point out there are technically other ways to boot a Live image, but given that there will be support for hiding arbitrary devices, that should not be a problem hopefully.

cockpituous commented on pkg/storaged/anaconda.jsx:
```diff
@@ -0,0 +1,42 @@
+    if (!client.in_anaconda_mode())
+        return null;
+
+    return (
+        <StackItem>
+            <Alert isInline
+                   variant='info'
+                   title={_("What you need to do")}>
+                <p>Anaconda will tell us here what is wrong with the current config.</p>
+            </Alert>
+        </StackItem>
```
These 8 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/block/format-dialog.jsx:
```diff
@@ -242,6 +247,10 @@ function format_dialog_internal(client, path, start, size, enable_dos_extended,
 
+    old_dir = client.strip_mount_point_prefix(old_dir);
+    if (old_dir === false)
+        return Promise.reject(_("This device can not be used for the installation target."));
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -741,6 +742,15 @@ function init_model(callback) {
 
+    try {
+        client.anaconda = JSON.parse(window.localStorage.getItem("cockpit_anaconda"));
+    } catch {
+        console.warn("Can't parse cockpit_anaconda configuration as JSON");
+        client.anaconda = null;
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+client.in_anaconda_mode = () => !!client.anaconda;
+
+client.strip_mount_point_prefix = (dir) => {
+    const mpp = client.anaconda?.mount_point_prefix;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+client.strip_mount_point_prefix = (dir) => {
+    const mpp = client.anaconda?.mount_point_prefix;
+
+    if (dir && mpp) {
+        if (dir.indexOf(mpp) != 0)
+            return false;
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+        if (dir.indexOf(mpp) != 0)
+            return false;
+
+        dir = dir.substr(mpp.length);
+        if (dir == "")
+            dir = "/";
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+};
+
+client.add_mount_point_prefix = (dir) => {
+    const mpp = client.anaconda?.mount_point_prefix;
+    if (mpp) {
+        if (dir == "/")
+            dir = mpp;
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+        if (dir == "/")
+            dir = mpp;
+        else
+            dir = mpp + dir;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+    return dir;
+};
+
+client.should_ignore_device = (devname) => {
+    return client.anaconda?.available_devices && client.anaconda.available_devices.indexOf(devname) == -1;
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1494,56 @@ client.get_config = (name, def) => {
+    return client.anaconda?.available_devices && client.anaconda.available_devices.indexOf(devname) == -1;
+};
+
+client.should_ignore_block = (block) => {
+    return client.should_ignore_device(utils.decode_filename(block.PreferredDevice));
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1106,7 +1106,8 @@ export const BlockingMessage = (usage) => {
         "mdraid-member": _("member of MDRAID device"),
         vdo: _("backing device for VDO device"),
-        "stratis-pool-member": _("member of Stratis pool")
+        "stratis-pool-member": _("member of Stratis pool"),
+        mounted: _("Filesystem outside the target"),
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1196,9 +1197,15 @@ export const TeardownMessage = (usage, expect_single_unmount) => {
+            if (use.usage == "mounted") {
+                location = client.strip_mount_point_prefix(location);
+                if (location === false)
+                    location = _("(Not part of target)");
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/filesystem/filesystem.jsx:
```diff
@@ -84,11 +84,13 @@ export function make_filesystem_card(next, backing_block, content_block, fstab_c
+    if (mount_point) {
+        mp_text = client.strip_mount_point_prefix(mount_point);
+        if (mp_text == false)
+            return null;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/filesystem/mounting-dialog.jsx:
```diff
@@ -47,6 +47,10 @@ export function mounting_dialog(client, block, mode, forced_options) {
 
+    const old_dir_for_display = client.strip_mount_point_prefix(old_dir);
+    if (old_dir_for_display === false)
+        return Promise.reject(_("This device can not be used for the installation target."));
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/stratis/pool.jsx:
```diff
@@ -106,7 +109,7 @@ function create_fs(pool) {
             SelectOne("at_boot", _("At boot"),
                       {
-                          value: "nofail",
+                          value: client.in_anaconda_mode() ? "local" : "nofail",
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/utils.js:
```diff
@@ -622,9 +624,13 @@ export function get_parent(client, path) {
         return client.blocks_stratis_fsys[path].Pool;
+    if (client.vgroups[path])
+        return path;
+    if (client.stratis_pools[path])
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/utils.js:
```diff
@@ -622,9 +624,13 @@ export function get_parent(client, path) {
+        return path;
+    if (client.stratis_pools[path])
+        return path;
 }
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/utils.js:
```diff
@@ -659,6 +665,19 @@ export function is_netdev(client, path) {
+        return false;
+
+    const parents = get_direct_parent_blocks(client, path);
+    if (parents.length == 0) {
+        const b = client.blocks[path];
+        return b && client.should_ignore_block(b);
+    } else {
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/utils.js:
```diff
@@ -659,6 +665,19 @@ export function is_netdev(client, path) {
+        return b && client.should_ignore_block(b);
+    } else {
+        return parents.some(p => should_ignore(client, p));
+    }
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-084155-675f6df4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -741,6 +741,15 @@ function init_model(callback) {
 
+    try {
+        client.anaconda = JSON.parse(window.localStorage.getItem("cockpit_anaconda"));
+    } catch {
+        console.warn("Can't parse cockpit_anaconda configuration as JSON");
+        client.anaconda = null;
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1493,40 @@
+
+    if (dir && mpp) {
+        if (dir.indexOf(mpp) != 0)
+            return false;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1493,40 @@
+    const mpp = client.anaconda?.mount_point_prefix;
+    if (mpp) {
+        if (dir == "/")
+            dir = mpp;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1493,40 @@
+        if (dir == "/")
+            dir = mpp;
+        else
+            dir = mpp + dir;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-145133-544b36e5-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1196,9 +1197,15 @@
+            if (use.usage == "mounted") {
+                location = client.strip_mount_point_prefix(location);
+                if (location === false)
+                    location = _("(Not part of target)");
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231205-120952-f0d04c46-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231211-144024-c4e15dfd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-092508-45201d04-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231212-131008-f8bec949-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented on doc/anaconda.md:
````diff
@@ -0,0 +1,70 @@
+Ignoring storage devices
+------------------------
+
+Anaconda needs to tell Cockpit which devices can not be used to
+install the OS on. This is done with the "available_devices" entry,
+which is an array of strings.
+
+```
+{
+  "available_devices": [ "/dev/sda" ]
+}
````
This is a bit misleading or confusing: That sounds like anaconda has to enumerate all available devices, substract the ones which can *not* be used to install, and send the rest? After the paragraph I had expected an `ignore_devices` instead, which might be easier to compute for Anaconda.

mvollmer replied:
We started out with "ignore_devices", but I switched it to "available_devices" to make things easier for Anaconda, actually. :-) It already has the list of useable devices, and there is a direct D-Bus call to get them, GetUseableDisks.

mvollmer replied:
> After the paragraph I had expected an ignore_devices instead.

Yeah, that's because it was written when we actually had "ignore_devices"...

martinpitt commented on doc/anaconda.md:
```diff
@@ -0,0 +1,70 @@
+omitted. So when the user says to create a filesystem on "/var", they
+are actually creating one on "/sysroot/var".
+
+However, Cockpit (via UDisks2) will still write the new mount point
+configuration into the real /etc/fstab (_not_ /sysroot/etc/fstab).
```
Is that the way how C-Storage "hands over" back to Anaconda? I.e. Anaconda will inspect /etc/fstab to see what the user has done? Or does it even have to care? At first sight it seems Anaconda can just assume "whatever the user did to /sysroot, I'll just install into it", but then again (1) boot, EFI etc. partitions are special and possibly not even mounted, and (2) Anaconda will still have to write /sysroot/etc/fstab, so I suppose it needs to handle these somehow.

This paragraph could explain in what way the written /etc/fstab is relevant.

mvollmer replied:
> Is that the way how C-Storage "hands over" back to Anaconda?

No. We write to fstab because this is the easiest thing to do for Cockpit. I have explained that now.

martinpitt replied:
I see -- "If and how Cockpit communicates back to Anaconda is still open." . So I take it Anaconda doesn't actually care that much, it'll just happily copy stuff into /sysroot and not second-guess that? But how does it create /sysroot/etc/fstab then? Anyway, that issue is blindingly obvious once Anaconda starts to actually use this, so this can/will happen in a follow-up and doesn't need to block this PR. Thanks for making it explicit in the docs!

martinpitt commented on pkg/storaged/anaconda.jsx:
```diff
@@ -0,0 +1,28 @@
+    if (!client.in_anaconda_mode())
+        return null;
+
+    // Nothing yet.
+    return null;
```
I'm fine with that for the first "secret" version, but what's the idea what this should eventually be? Some kind of inline help for what admins are supposed to do on the page, and how to go back?

martinpitt replied:
I leave that up to you, just wanted to point it out.

martinpitt commented on pkg/storaged/stratis/filesystem.jsx:
```diff
@@ -19,7 +19,7 @@
 import cockpit from "cockpit";
 import React from "react";
-import client from "../client";
+import client from "../client.js";
```
Mausgerutscht? All other pages import it without extension.

martinpitt commented on test/verify/check-storage-anaconda:
```diff
@@ -0,0 +1,121 @@
+        self.dialog_apply()
+        self.dialog_wait_close()
+
+        self.assertNotIn("noauto", m.execute("findmnt --fstab -n -o OPTIONS /sysroot"))
```
This should test the absence of `nofail` as well I think?

martinpitt replied:
I thought it should be in c-s-anaconda because the behaviour is now conditional. The non-anaconda tests should indeed check for `nofail` to be present, and they do (e.g. test/verify/check-storage-mounting).

martinpitt commented on test/verify/check-storage-anaconda:
```diff
@@ -0,0 +1,121 @@
+        # Back to the beginning
+        b.wait_visible(self.card("Storage"))
+        b.wait(lambda: b.call_js_func('ph_count', self.card("Storage") + " tbody tr") == 1)
+        b.wait_not_present(self.card_row("Storage", location="/"))
```
There are a lot of special cases/code changes in the stratis code. Could this be covered as well here? It needs a loop device instead of a RAM disk, but you can either replace it or just add a second one.

martinpitt reviewed (changes requested):
Thanks! This is quite intrusive, but I suppose that's rather unavoidable. The code changes LGTM, I have some questions/remarks about the design and test coverage.

mvollmer commented on pkg/storaged/nfs/nfs.jsx:
```diff
@@ -280,6 +280,9 @@ const NfsEntryUsageBar = ({ entry, not_mounted_text, short }) => {
 };
 
 export function make_nfs_page(parent, entry) {
+    if (client.in_anaconda_mode())
+        return;
```
This is indeed the idea: No NFS in anaconda mode. The code implements this.

mvollmer replied:
But I am not so sure about iSCSI... It doesn't make much sense to configure "boot fom NFS" in Cockpit Storage, but iSCSI is just block devices that Anaconda can untar the OS onto... maybe that's a thing.

martinpitt replied:
Ack thanks, so I understood that correctly.

martinpitt commented on test/verify/check-storage-anaconda:
```diff
@@ -0,0 +1,201 @@
+        b.reload()
+        b.enter_page("/storage")
+
+        testlib.sit()
```
Standup time! :clock3: :grin:

mvollmer replied:
haha!

martinpitt commented on test/verify/check-storage-anaconda:
```diff
@@ -0,0 +1,201 @@
+        self.dialog_wait_close()
+
+        self.assertNotIn("noauto", m.execute("findmnt --fstab -n -o OPTIONS /sysroot"))
+        self.assertNotIn("nofail", m.execute("findmnt --fstab -n -o OPTIONS /sysroot"))
```
`findmnt` will fail if the mount doesn't exist, but it would still feel a bit better to also have some positive test here. Maybe that it contains `lvol0` and `defaults` or `auto`?

mvollmer replied:
We have a positive test further down, is that enough?

mvollmer replied:
There isn't really anything interesting in options...

martinpitt replied:
Right, but it can also show SOURCE ?

martinpitt replied:
> We have a positive test further down, is that enough?

I guess.. I don't want to dwell on this, I leave this up to you. TBH, my gut feeling is that this is by far not the end of the whole Anaconda integration story, we need to have some much more conceptual discussions. Hence, I'll almost surely approve your next force-push :grin:

mvollmer replied:
> TBH, my gut feeling is that this is by far not the end of the whole Anaconda integration story,

Of course not. :-)  The current goal is to get something into Cockpit main that Anaconda can test against.

mvollmer replied:
> I guess.. I don't want to dwell on this, I leave this up to you.

Ok, let's leave it as it is. As you say, findmnt will fail noisily when the expected /sysroot entry is not there.

martinpitt commented on test/verify/check-storage-anaconda:
```diff
@@ -0,0 +1,201 @@
+        self.dialog_apply()
+        self.dialog_wait_close()
+
+        self.assertNotIn("noauto", m.execute("findmnt --fstab -n -o OPTIONS /sysroot"))
+        self.assertNotIn("nofail", m.execute("findmnt --fstab -n -o OPTIONS /sysroot"))
```
Dito about a positive check

martinpitt reviewed (changes requested):
Some remaining tiny nits, but from my POV this is basically ready to go. Then @KKoukiou can start playing around with it from the main-builds COPR.

martinpitt reviewed (dismissed):
:rocket: as this remains fully malleable for the time being.

mvollmer commented:
Upps, I have found a bug: leaving the "Mount point" field empty results in "/" being used as the mount point. Instead it should keep the meaning of "no mount point specified" and either be an error, or lead to skipping of fstab updates.

mvollmer commented:
Changes since last review:

```
client.add_mount_point_prefix = (dir) => {
    const mpp = client.anaconda?.mount_point_prefix;
-    if (mpp) {
+    if (mpp && dir != "") {
        if (dir == "/")
            dir = mpp;
        else
            dir = mpp + dir;
    }
    return dir;
};
```

And the bug fix in [storage: Only action failures should disable the apply button](https://github.com/cockpit-project/cockpit/pull/19352/commits/4195ba3a4c624c8659ad61dad9b6cd5c370f1491)

mvollmer commented:
Pixels need to updated.

martinpitt reviewed (approved):
i386 unit test has some trouble, retried.

mvollmer commented on pkg/storaged/client.js:
```diff
@@ -741,6 +741,15 @@ function init_model(callback) {
+        client.anaconda = null;
+    }
+
+    console.log("ANACONDA", client.anaconda);
```
Hmm, this runs unconditionally... it's better to remove it I guess.

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1493,40 @@
+    const mpp = client.anaconda?.mount_point_prefix;
+    if (mpp && dir != "") {
+        if (dir == "/")
+            dir = mpp;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/client.js:
```diff
@@ -1484,4 +1493,40 @@
+        if (dir == "/")
+            dir = mpp;
+        else
+            dir = mpp + dir;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1106,7 +1107,8 @@ export const BlockingMessage = (usage) => {
         "mdraid-member": _("member of MDRAID device"),
         vdo: _("backing device for VDO device"),
-        "stratis-pool-member": _("member of Stratis pool")
+        "stratis-pool-member": _("member of Stratis pool"),
+        mounted: _("Filesystem outside the target"),
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/dialog.jsx:
```diff
@@ -1196,9 +1198,15 @@
+            if (use.usage == "mounted") {
+                location = client.strip_mount_point_prefix(location);
+                if (location === false)
+                    location = _("(Not part of target)");
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20231213-115423-cbae7212-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/storaged/stratis/pool.jsx:
```diff
@@ -107,7 +110,7 @@ function create_fs(pool) {
             SelectOne("at_boot", _("At boot"),
                       {
-                          value: "nofail",
+                          value: client.in_anaconda_mode() ? "local" : "nofail",
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19352-20240102-111247-0a02f8c6-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)
